src/main/AllocationTable.cpp

```diff
@@ -0,0 +1,286 @@
+ *  Copyright (C) 2008-2014  Andrew R. Runnalls.
+ *  Copyright (C) 2014 and onwards the Rho Project Authors.
+ *
+ *  Rho is not part of the R project, and bugs and other issues should
+ *  not be reported via r-bugs or other R project channels; instead refer
+ *  to the Rho website.
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, a copy is available at
+ *  https://www.R-project.org/Licenses/
+ */
+
+#include <assert.h>
+#include <stdio.h>
+
+#include <cstdint>
+#include <functional>
+#include <limits>
+
+#include "rho/AddressSanitizer.hpp"
+#include "rho/AllocationTable.hpp"
+#include "rho/GCNodeAllocator.hpp"
+
+rho::AllocationTable::AllocationTable(unsigned num_bits): m_num_bits(num_bits) {
+  m_num_buckets = 1 << m_num_bits;
+  m_capacity = m_num_buckets / 2;  // Sets max load factor to 50%.
+  m_hash_mask = m_num_buckets - 1;
+  m_buckets = new Allocation[m_num_buckets];
+}
+
+rho::AllocationTable::~AllocationTable() {
+  delete[] m_buckets;
+  m_buckets = nullptr;
+}
+
+void rho::AllocationTable::insert(void* pointer, std::size_t size) {
+  if (m_capacity == 0) {
+    if (!resize(m_num_bits + 1)) {
+      allocerr("failed to increase allocation table size");
+    }
+  }
+  m_capacity -= 1;
+  if (!tryInsert(m_buckets, reinterpret_cast<uintptr_t>(pointer), size)) {
+    allocerr("failed to insert allocation into allocation table");
+  }
+}
+
+void* rho::AllocationTable::lookup_pointer(uintptr_t candidate) {
+  Allocation* bucket = search(candidate);
+  if (bucket) {
+    if (bucket->isSuperblock()) {
+      AllocatorSuperblock* superblock = bucket->asSuperblock();
+      return superblock->lookupBlock(candidate);
+    } else {
+      return bucket->asPointer();
+    }
+  }
+  return nullptr;
+}
+
+void rho::AllocationTable::erase(uintptr_t pointer, size_t size) {
+  pointer &= ~uintptr_t{3};  // Mask out the flag bits for equality comparison.
+  size_t key_end = endKey(pointer, size);
+  bool first = true;
+  for (size_t key = startKey(pointer); key < key_end; ++key) {
+    size_t hash = pointerHash(key << LOW_BITS);
+    for (int i = 0; i < m_num_buckets; ++i) {
+      Allocation& bucket = m_buckets[hash];
+      if (bucket.dataPointer() == pointer) {
+        m_capacity += 1;
+        bucket.m_data = Allocation::s_deleted_key;
+        break;
+      }
+      if (bucket.isEmpty()) {
+        // No more partial entries inserted, we are done.
```

Comment:
Yes, this behavior was left over from when I expected every insert to be able to fail, and erase was used to clean up partially inserted keys. Made it more sane now so that it generates an abort().